library/alloc/tests/vec.rs

```diff
@@ -287,6 +287,78 @@ fn test_retain() {
     assert_eq!(vec, [2, 4]);
 }
 
+#[test]
+fn test_retain_pred_panic() {
+    use std::sync::atomic::{AtomicU64, Ordering};
```

Comment:
I'm not sure if we can assume AtomicU64 is available on all platforms these tests will run on. Maybe this can use someting with an array of `Cell<bool>` instead of atomics?

<details><summary>Example</summary>

```rust
#[test]
fn test_retain_pred_panic() {
    use std::cell::Cell;

    let dropped: [Cell<bool>; 5] = Default::default();

    struct Wrap<'a>(&'a Cell<bool>, usize);

    impl Drop for Wrap<'_> {
        fn drop(&mut self) {
            self.0.set(true);
        }
    }

    let ret = std::panic::catch_unwind(AssertUnwindSafe(|| {
        let mut v = vec![
            Wrap(&dropped[0], 0),
            Wrap(&dropped[1], 1),
            Wrap(&dropped[2], 2),
            Wrap(&dropped[3], 3),
            Wrap(&dropped[4], 4),
        ];
        v.retain(|w| match w.1 {
            0 => true,
            1 => false,
            2 => true,
            _ => panic!(),
        });
    }));
    assert!(ret.is_err());
    // Everything is dropped when predicate panicked.
    assert!(dropped.iter().all(|d| d.get()));
}
```
</details>

Another option is to use `Rc` or `Arc`, put a clone in the vec, and check at the end of the count on the original one has dropped back to 1. Then you don't need a custom type or Drop implementation.